content/post/71-putting-npm-on-ipfs-part-2.md

```diff
@@ -0,0 +1,82 @@
+
+Using the registry in this way works, but it still depends on HTTP and DNS to resolve your dependencies. So far so centralised.
+
+What if there was a way to skip that bit and have it be distributed-web-turtles all the way down?
+
+** drumroll **
+
+## 🙌 Introducing ipfs-npm
+
+[`ipfs-npm`](https://www.npmjs.com/package/ipfs-npm) is a command line tool that spins up an IPFS node and uses that to resolve the tarballs that make up your project's dependencies ([other configurations are available](https://www.npmjs.com/package/ipfs-npm#cli)). It works with npm and Yarn, and will add your dependencies to the [IPFS Repository](https://github.com/ipfs/specs/tree/master/repo) of your node so you can resolve them offline later on.
+
+### 🚚 Installation
+
+You can install it with:
+
+```console
+$ npm install -g ipfs-npm
+```
+
+### 🔧 Usage
+
+`ipfs-npm` proxies for the npm cli, passing all arguments on, so use it as you would npm, just with a different command name:
+
+```console
+$ cd my-project
+$ ipfs-npm install
+👿 Spawning an in-process IPFS node using repo at /Users/alex/.jsipfs
+Swarm listening on /ip4/127.0.0.1/tcp/57029/ipfs/QmZ7vEvXRdTVipb9o2p2Cmt3s1S8rqbo2ohscTjvTLgnpP
+🗂️ Loading registry index from https://registry.js.ipfs.io
+☎️ Dialling registry mirror /ip4/35.178.192.119/tcp/10015/ipfs/QmWBaYSnmgZi6F6D69JuZGhyL8rm6pt8GX5r7Atc6Gd7vR,/dns4/registry.js.ipfs.io/tcp/10015/ipfs/QmWBaYSnmgZi6F6D69JuZGhyL8rm6pt8GX5r7Atc6Gd7vR
+📱️ Connected to registry
+👩‍🚀 Starting local proxy
+🚀 Server running on port 37847
+🎁 Installing dependencies with /Users/alex/.nvm/versions/node/v10.15.1/bin/npm
+...
+```
+
+If you prefer Yarn and have it installed globally, use the `ipfs-yarn` command (installed alongside `ipfs-npm`):
+
+```console
+$ cd my-project
+$ ipfs-yarn
+👿 Spawning an in-process IPFS node using repo at /Users/alex/.jsipfs
+Swarm listening on /ip4/127.0.0.1/tcp/49905/ipfs/QmZ7vEvXRdTVipb9o2p2Cmt3s1S8rqbo2ohscTjvTLgnpP
+🗂️  Loading registry index from https://registry.js.ipfs.io
+☎️  Dialling registry mirror /ip4/35.178.192.119/tcp/10040/ipfs/QmfKqxieE71QoAchNk5e2MKmvWKjGdUnSifHqq1xZLEzyn,/dns4/registry.js.ipfs.io/tcp/10040/ipfs/QmfKqxieE71QoAchNk5e2MKmvWKjGdUnSifHqq1xZLEzyn
+📱️ Connected to registry
+👩‍🚀 Starting local proxy
+🚀 Server running on port 49935
+🎁 Installing dependencies with /Users/alex/.nvm/versions/node/v10.15.1/bin/yarn
+yarn install v1.13.0
+...
+```
+
+## 🙋 How does it work?
+
+Behind the scenes, it starts a local http server and configures npm/Yarn to use it, it then uses the same package metadata as [registry.js.ipfs.io](https://registry.js.ipfs.io) to resolve a CID for a requested tarball - if it has the CID, it uses IPFS to request the content, otherwise it downloads it from the public npm registry, adds it to IPFS and stores the CID for the next time it's requested.
+
+![Request sequence](/71-putting-npm-on-ipfs-part-2/ipfs-npm-sequence.png)
```

Comment:
img doesn't load for me

Reply:
@achingbrain can you host the image on IPFS instead?

Reply:
@momack2 Weird, it works for me - do the other images load?

@daviddias should the image CID be pinned to the cluster or similar?  How do you go about that?

Reply:
@achingbrain for one-offs like this you can talk to ipfspin on twitter or ipfspinbot on slack

Reply:
for example https://twitter.com/ipfspin/status/1103063297862103041

Reply:
the entire site is built out and the root dir is pinned to our cluster (or at least, it will be when this is merged to master) so all the content will be bundled under the new IPFS CID for the site root.

The image should load just fine, and doesn't need to be pinned to IPFS separately. It'd be worth double checking that hugo is auto-relativising the img src url, otherwise it'll load correctly when the site is mounted at the domain root (blog.ipfs.io) but won't when it's mounted under ipfs (ipfs.io/ipfs/bafyhashy)

Reply:
I'm gonna assume this will work when pushed live. =]